addr_manager_ds.go

```diff
@@ -0,0 +1,316 @@
+	hash, err := mh.Sum((*addr).Bytes(), mh.MURMUR3, -1)
+	if err != nil {
+		return ds.Key{}, nil
+	}
+	return ds.NewKey(peer.IDB58Encode(*p)).ChildString(hash.B58String()), nil
+}
+
+func peerIDFromKey(key ds.Key) (peer.ID, error) {
+	idstring := key.Parent().Name()
+	return peer.IDB58Decode(idstring)
+}
+
+// AddAddr will add a new address if it's not already in the AddrBook.
+func (mgr *DatastoreAddrManager) AddAddr(p peer.ID, addr ma.Multiaddr, ttl time.Duration) {
+	mgr.AddAddrs(p, []ma.Multiaddr{addr}, ttl)
+}
+
+// AddAddrs will add many new addresses if they're not already in the AddrBook.
+func (mgr *DatastoreAddrManager) AddAddrs(p peer.ID, addrs []ma.Multiaddr, ttl time.Duration) {
+	if ttl <= 0 {
+		return
+	}
+
+	mgr.setAddrs(p, addrs, ttl, true)
+}
+
+// SetAddr will add or update the TTL of an address in the AddrBook.
+func (mgr *DatastoreAddrManager) SetAddr(p peer.ID, addr ma.Multiaddr, ttl time.Duration) {
+	mgr.SetAddrs(p, []ma.Multiaddr{addr}, ttl)
+}
+
+// SetAddrs will add or update the TTLs of addresses in the AddrBook.
+func (mgr *DatastoreAddrManager) SetAddrs(p peer.ID, addrs []ma.Multiaddr, ttl time.Duration) {
+	mgr.setAddrs(p, addrs, ttl, false)
+}
+
+func (mgr *DatastoreAddrManager) setAddrs(p peer.ID, addrs []ma.Multiaddr, ttl time.Duration, add bool) {
+	for i := 0; i < dsWriteRetries; i++ {
+		// keys to add to the TTL manager
+		var keys []ds.Key
+		batch, err := mgr.ds.Batch()
+		if err != nil {
+			log.Error(err)
+			return
+		}
+
+		for _, addr := range addrs {
+			if addr == nil {
+				continue
+			}
+
+			key, err := peerAddressKey(&p, &addr)
+			if err != nil {
+				log.Error(err)
+				continue
+			}
+			keys = append(keys, key)
+
+			if ttl <= 0 {
+				batch.Delete(key)
+				continue
+			}
+			has, err := mgr.ds.Has(key)
+			if err != nil || !has {
+				mgr.subsManager.BroadcastAddr(p, addr)
+			}
+
+			// Allows us to support AddAddr and SetAddr in one function
+			if !has || !add {
+				if err := batch.Put(key, addr.Bytes()); err != nil {
+					log.Error(err)
+				}
+			}
+		}
+		if err := batch.Commit(); err != nil {
+			log.Errorf("failed to write addresses for peer %s: %s\n", p.Pretty(), err)
+			continue
+		}
+		mgr.ttlManager.setTTLs(keys, ttl)
+		return
+	}
+	log.Errorf("failed to avoid write conflict for peer %s after %d retries\n", p.Pretty(), dsWriteRetries)
+}
+
+// UpdateAddrs will update any addresses for a given peer and TTL combination to
+// have a new TTL.
+func (mgr *DatastoreAddrManager) UpdateAddrs(p peer.ID, oldTTL time.Duration, newTTL time.Duration) {
+	prefix := ds.NewKey(p.Pretty())
+	mgr.ttlManager.updateTTLs(prefix, oldTTL, newTTL)
+}
+
+// Addrs Returns all of the non-expired addresses for a given peer.
+func (mgr *DatastoreAddrManager) Addrs(p peer.ID) []ma.Multiaddr {
+	prefix := ds.NewKey(p.Pretty())
+	q := query.Query{Prefix: prefix.String()}
+	results, err := mgr.ds.Query(q)
+	if err != nil {
+		log.Error(err)
+		return []ma.Multiaddr{}
```

Comment:
`nil` is the preferred empty value for arrays, it behaves properly in all cases (acts just like a zero length array, len() == 0, and ranging over it is a noop) and is just makes the intent a bit clearer.